### PR TITLE
[WIP] `/model` endpoint

### PR DESCRIPTION
**Pull request by stefanorosanelli**
`/model` endpoints documentation

Still missing:
 - `/model/properties` - will be updated in this PR
 - `/model/property_types` - in a future PR - relevant code changes are expected

### CODE REVIEW

File: source/endpoints/model/object_types.rst [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+                "description": "This is a cat"
+        }
+    }
+
+In its simplest form the only required attributes are:
+
+* ``name`` object type in plural form, this is the canonical form for this type that will activate ``/cats`` endpoint
+* ``singular`` singular form for type
+
+Both must be in `lowered snake_case <https://en.wikipedia.org/wiki/Snake_case>`_ format as internal naming convention, see :ref:`api-model-naming`, and must contain at least a letter (a-z) to avoid collusions with numeric id.
```
Comment (didoda) [inline source/endpoints/model/object_types.rst]: `collisions`?

File: source/endpoints/model/object_types.rst [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+                }
+            }
+        },
+        "links": {
+            "self": "http://api.example.com/model/object_types/cats",
+            "home": "http://api.example.com/home"
+        }
+    }
+
+Example reponse shows us some additional interesting fields:
```
Comment (didoda) [inline source/endpoints/model/object_types.rst]: typo: `response`

File: source/endpoints/model/object_types.rst [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+
+You can modify an object type by using ``PATCH /model/object_types/{type}`` endpoint.
+
+.. http:patch:: /model/object_types/{type}
+
+* ``{type}`` can be a numeric id of the object type or its ``name``
+
+**Example request: modify an object type**:
+
+In this example we will simply disable the newly created type and chage its description
```
Comment (didoda) [inline source/endpoints/model/object_types.rst]: typo: `change`

File: source/endpoints/model/object_types.rst [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+``"meta"`` and ``"relationships"`` here omitted for brevity.
+
+Remove an object type
+---------------------
+
+You can delete permanently an object type by using ``DELETE /model/object_types/{type}`` endpoint.
+
+This operation cannot be reversed and will not be allowed if:
+
+* object of this type exist
```
Comment (didoda) [inline source/endpoints/model/object_types.rst]: `exists` or reformulate as `one or more objects of this type exist`

File: source/endpoints/model/object_types.rst [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+**Example request: delete cats**:
+
+.. sourcecode:: http
+
+    DELETE /model/object_types/cats HTTP/1.1
+    Host: api.example.com
+
+Expected HTTP status response is ``204 No Content``.
+
+If object type is not not found, response will be ``404 Not Found``, if delete operation is not allowed a ``403 Forbidden`` will be sent.
```
Comment (didoda) [inline source/endpoints/model/object_types.rst]: typo: not ~not~

File: source/endpoints/model/properties.rst [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+.. sourcecode:: http
+
+    GET /model/object_types?filter[object_type]=documents HTTP/1.1
+    Accept: application/vnd.api+json
+
+Response will contain an array of ``properties`` in typical list format as shown in :ref:`api-responses` belonging to ``documents`` type only
+
+In this response you will note two important things:
+
+* some items contain an hash format id like ``"id": "608e814b-e4d7-57c2-8599-9692803e30bc"`` and some a classic numeric one; has format _ids_   represent ``static`` fields, i.e. fields that match actual database columns that in general cannot be removed, but can be hidden - see ``hidden`` object types attribute in :ref:`api-model-object-types`; instead pure numeric _ids_ represent pure ``dynamic`` properties defined at project level; ``static`` will be present in any project, ``dynamic`` are custom ones available on the current project only.
```
Comment (didoda) [inline source/endpoints/model/properties.rst]: typo: `hash` (in `; hash format ...`)

File: source/endpoints/model/properties.rst [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+
+To change a property there is a ``PATCH /model/properties/{{id}}`` method available.
+
+Remember that only ``dynamic`` properties may be modified, whereas ``static`` ones are immutable, but can be hidden.
+
+.. http:patch:: /model/properties/{{id}}
+
+* ``{id}`` dynamic property numeric ``id``
+
+**Example request: modify an object type**:
```
Comment (didoda) [inline source/endpoints/model/properties.rst]: typo: `modify a property`

File: source/endpoints/model/properties.rst [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+
+In this example we will simply disable the newly created type and chage its description
```
Comment (didoda) [inline source/endpoints/model/properties.rst]: typos: `newly created property` / `change`

File: source/endpoints/model/properties.rst [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+
+**Example request: delete property**:
+
+.. sourcecode:: http
+
+    DELETE /model/properties/1 HTTP/1.1
+
+Expected HTTP status response is ``204 No Content``.
+
+If property is not not found, response will be ``404 Not Found``, if delete operation is not allowed a ``403 Forbidden`` will be sent.
```
Comment (didoda) [inline source/endpoints/model/properties.rst]: typo: not ~not~

File: source/endpoints/model/relations.rst [hunk 10]
```diff
@@ -0,0 +1,351 @@
+Relations ``/model/relations``
+==============================
+
+``/model/relations`` endpoint manages relations between objects types.
```
Comment (didoda) [inline source/endpoints/model/relations.rst]: typo: `object types`

File: source/endpoints/model/relations.rst [hunk 11]
```diff
@@ -0,0 +1,351 @@
...
+
+This is one of the most important features in BEdita: you can create dynamic semantic relations between object types, adding constraint rules and parameters without any data model or database modification.
+
+Relations in BEdita have been heavily inspired by semantic web `RDF triples <https://www.w3.org/TR/2014/NOTE-rdf11-primer-20140624/#section-triple>`_.
+
+Every relation should be espressed in a form ``{subject} {predicate} {object}`` where both ``{subject}`` and ``{object}`` are BEdita objects and ``{predicate}`` is the relation name.
```
Comment (didoda) [inline source/endpoints/model/relations.rst]: typo: `expressed`

File: source/endpoints/model/relations.rst [hunk 12]
```diff
@@ -0,0 +1,351 @@
...
+With this endpoint you can:
+
+* create, update/remove an object relation
+* list available relations
+* handle which object types are involved in a relation
+
+Create a relation
+-----------------
+
+Creation of a new object type happens through ``POST /model/relations`` endpoint.
```
Comment (didoda) [inline source/endpoints/model/relations.rst]: `Creation of a new relation`

File: source/endpoints/model/relations.rst [hunk 13]
```diff
@@ -0,0 +1,351 @@
...
+Required attributes creating a new relation are:
+
+* ``name`` principal relation name
+* ``inverse_name`` inverse relation name
+
+Both attributes must be different and unique inside the project, should differ from property names of object types involved in this relation and as naming convention should also be expressed as a _predicate_ linking object types together.
+
+They also must be in `lowered snake_case <https://en.wikipedia.org/wiki/Snake_case>`_ format as internal naming convention.
+
+Our goal in this example, altough not yet complete, is to create the relation ``[users] [owner_of] [cats]`` and its inverse ``[cats] [belong_to] [users]``.
```
Comment (didoda) [inline source/endpoints/model/relations.rst]: typo: `although`

File: source/endpoints/model/relations.rst [hunk 14]
```diff
@@ -0,0 +1,351 @@
...
+Expected response is ``HTTP/1.1 201 OK``, with ``application/vnd.api+json`` body data representing relation just created.
+
+If the relation already exists or data is not valid (i.e. lacks of required fields or inconsistent input), POST fails and response will be ``400 Bad Request - Invalid data``.
+
+.. _api-model-relations-get:
+
+Get a single relation
+---------------------
+
+You can obtain a single relations by invoking ``GET /model/relations/{relation}``.
```
Comment (didoda) [inline source/endpoints/model/relations.rst]: typo: `single relation`

File: source/endpoints/model/relations.rst [hunk 15]
```diff
@@ -0,0 +1,351 @@
...
+                }
+            }
+        },
+        "links": {
+            "self": "http://api.example.com/model/relations/owner_of",
+            "home": "http://api.example.com/home"
+        }
+    }
+
+Example reponse shows us something interesting in ``"relationships"`` section:
```
Comment (didoda) [inline source/endpoints/model/relations.rst]: typo: `response`

File: source/endpoints/model/relations.rst [hunk 16]
```diff
@@ -0,0 +1,351 @@
...
+            {
+                "type": "object_types",
+                "id": "2"
+            }
+        ]
+    }
+
+In this example ``users`` (with _id_ 3) and ``profiles`` (_id_ 2) have been added as left side object types to the relation created before.
+
+Expected response is ``200 OK`` upon success, while response will be contain only a ``"links"`` section pointing to the list of left/right object types for the current relation.
```
Comment (didoda) [inline source/endpoints/model/relations.rst]: typo: response will ~be~ contain`

File: source/endpoints/model/relations.rst [hunk 17]
```diff
@@ -0,0 +1,351 @@
...
+    {
+        "data": [
+            {
+                "type": "object_types",
+                "id": "13"
+            }
+        ]
+    }
+
+With this method we have completed our relation putting our custom object type seen in :ref:`api-model-object-types` on the righ side.
```
Comment (didoda) [inline source/endpoints/model/relations.rst]: typo: `right side`

File: source/endpoints/model/relations.rst [hunk 18]
```diff
@@ -0,0 +1,351 @@
...
+
+Now we have: ``{users | profiles} {owner_of} {cats}`` and inverse from right to left ``{cats} {belong_to} {users | profiles}``.
+
+Expected response is ``200 OK`` upon success, while response will be contain only a ``"links"`` section pointing to the list of left/right object types for the current relation.
```
Comment (didoda) [inline source/endpoints/model/relations.rst]: Drop this line? it seems not necessary

File: source/endpoints/model/relations.rst [hunk 19]
```diff
@@ -0,0 +1,351 @@
...
+
+You can modify a relation by using ``PATCH /model/relations/{relation}`` endpoint.
+
+.. http:patch:: /model/relations/{relation}
+
+* ``{relation}`` can be a numeric id of the relation, its ``name`` or ``inverse_name``
+
+**Example request: modify a relation**:
+
+In this example we will simply disable the newly created type and chage its description
```
Comment (didoda) [inline source/endpoints/model/relations.rst]: `In this example we will just change the description for the relation 1`

File: source/endpoints/model/relations.rst [hunk 20]
```diff
@@ -0,0 +1,351 @@
...
+**Example request: delete relation**:
+
+.. sourcecode:: http
+
+    DELETE /model/relations/owner_of HTTP/1.1
+    Host: api.example.com
+
+Expected HTTP status response is ``204 No Content``.
+
+If relation is not not found, response will be ``404 Not Found``, if delete operation is not allowed a ``403 Forbidden`` will be sent.
```
Comment (didoda) [inline source/endpoints/model/relations.rst]: typo: not ~not~

File: source/endpoints/model/schema.rst [hunk 21]
```diff
@@ -0,0 +1,83 @@
...
+
+.. http:get:: /model/schema/{type}
+
+* ``{type}`` object type or resource ``name``
+
+**Example request: role schema**:
+
+.. sourcecode:: http
+
+    GET /model/object_types/roles HTTP/1.1
```
Comment (didoda) [inline source/endpoints/model/schema.rst]: maybe `GET /model/schema/roles HTTP/1.1`?

**Comment by stefanorosanelli** [comment]
Thanks @didoda all typos should be fixed now

**Pull request merged**